include/dali/core/mm/detail/free_list.h

```diff
@@ -323,6 +349,185 @@ class coalescing_free_list : public best_fit_free_list {
       *pwhere = blk;
     }
   }
+
+  void merge(coalescing_free_list &&with) {
+    if (!with.head_)
+      return;
+    if (!head_) {
+      swap(with);
+      return;
+    }
+
+    block **a = &head_;
+    block **b = &with.head_;
+    block *new_head = nullptr;
+
+    auto next_block = [&]() {
+      return (*a)->start < (*b)->start ? a : b;
+    };
+
+    block **src = next_block();
+    new_head = *src;
+    *src = (*src)->next;
+    new_head->next = nullptr;
+    block *tail = new_head;
+    while (*a && *b) {
+      src = next_block();
+      block *curr = *src;
+      *src = curr->next;  // advance the source
+      curr->next = nullptr;
+
+      if (curr->start == tail->end) {
+        // coalesce
+        tail->end = curr->end;
+        // Current block is no longer needed - place it in unused blocks in source list.
+        // This avoids growing the destination's unused blocks list indefinitely.
+        curr->next = with.unused_blocks_;
+        curr->start = curr->end = nullptr;
+        with.unused_blocks_ = curr;
+      } else {
+        // attach current element to the resuiting list
+        tail->next = curr;
+        // move tail
+        tail = curr;
+      }
+    }
+    // append whatever's left
+    for (block **p : { a, b }) {
+      if (*p) {
+        if ((*p)->start == tail->end) {
+          // coalesce
+          tail->end = (*p)->end;
+          with.remove(p);
```

Comment:
That's intentional. The target scenario is:
- a global list
- a bunch of thread-local lists.
When a thread terminates (or the resource is destroyed) the thread-local list would be merged into the global list. If `merge` would place items from the list about to be destroyed in global unused_blocks, it could grow indefinitely. True, this makes it more likely that we'll need to allocate list entries, but at least protects us from unbounded growth.

Reply:
I think I still miss the difference between `with.remove(p);` and `with.unused_blocks_ = curr;`.

Reply:
Well, remove does place it in the unused blocks - the point is that we place it in the unused blocks of `with` and not the unused blocks of `this`.

Reply:
So, with.remove(p) does
(all the boilerplate of unlinking p)
and places it at the head of with.unused_blocks_.